src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -0,0 +1,1503 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A class that serializes objects to a stream.
+    /// </summary>
+    internal sealed partial class StreamObjectWriter : ObjectWriter, IDisposable
+    {
+        private readonly BinaryWriter _writer;          // writes primitives to stream
+        private readonly ObjectBinder _binder;          // provides object and type encoding
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly ReferenceMap _referenceMap;        // object instance to reference-id map
+        private readonly Stack<Variant> _valueStack;        // stack of member/element values from object/array contents pending emit
+        private readonly VariantListWriter _memberWriter;   // used to get object member values
+
+        /// <summary>
+        /// Creates a new instance of a <see cref="StreamObjectWriter"/>.
+        /// </summary>
+        /// <param name="stream">The stream to write to.</param>
+        /// <param name="knownObjects">An optional list of objects assumed known by the corresponding <see cref="StreamObjectReader"/>.</param>
+        /// <param name="binder">A binder that provides object and type encoding.</param>
+        /// <param name="cancellationToken"></param>
+        public StreamObjectWriter(
+            Stream stream,
+            ObjectData knownObjects = null,
+            ObjectBinder binder = null,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // String serialization assumes both reader and writer to be of the same endianness.
+            // It can be adjusted for BigEndian if needed.
+            Debug.Assert(BitConverter.IsLittleEndian);
+
+            _writer = new BinaryWriter(stream, Encoding.UTF8);
+            _referenceMap = ReferenceMap.Create(knownObjects);
+            _binder = binder ?? new SimpleRecordingObjectBinder();
+            _cancellationToken = cancellationToken;
+            _valueStack = new Stack<Variant>();
```

Comment:
Yes, its possible to pool this stack and the member writer.

---
In reply to: [84802425](https://github.com/dotnet/roslyn/pull/14671#discussion_r84802425) [](ancestors = 84802425)